sys/src/cmd/gdbserver/gdbstub.c

```diff
@@ -134,18 +137,125 @@ p64(char *dest, uint64_t c)
 }
 
 void
+read_pid_text(int pid)
+{
+	char buf[100];
+	sprint(buf, "/proc/%d/text", pid);
+
+	// TODO close textfid?
+	int textfid = open(buf, OREAD);
+	if (textfid < 0){
+		syslog(0, "gdbserver", "couldn't open %s: %r", buf);
+		return;
+	}
+
+	Fhdr fhdr;
+	if (!crackhdr(textfid, &fhdr)) {
+		syslog(0, "gdbserver", "couldn't decode file header: %r");
+		return;
+	}
+
+	Map* symmap = loadmap(0, textfid, &fhdr);
+	if (symmap == 0) {
+		syslog(0, "gdbserver", "loadmap failed: %r");
+		return;
+	}
+
+	if (syminit(textfid, &fhdr) < 0) {
+		syslog(0, "gdbserver", "syminit failed: %r");
+		return;
+	}
+
+	machbytype(fhdr.type);
+
+	// TODO close memfid
+	snprint(buf, sizeof(buf), "/proc/%d/mem", pid);
```

Comment:
I think I'll do that - and for the ctl fid too, rather than open and close repeatedly.